R/vsc_r_api.R

```diff
@@ -0,0 +1,290 @@
+    # the console was the current selection, so it is not very useful.
+    editor_context <- request_response("active_editor_context")
+
+    make_rs_document_context(editor_context)
+}
+
+getSourceEditorContext <- getActiveDocumentContext
+
+verifyAvailable <- function(version_needed = NULL) {
+    if (is.null(version_needed)) TRUE else FALSE
+}
+
+isAvailable <- function(version_needed = NULL, child_ok) {
+    verify_available(version_needed)
+}
+
+insertText <- function(location, text, id = NULL) {
+
+    ## insertText also supports insertText("text"), insertText(text = "text"),
+    ## allowing the location parameter to be used for the text when
+    ## text itself is null.
+    ## This is dispatched as a separate request type
+    if (missing(text) && is.character(location) && length(location) == 1) {
+        ## handling insertText("text")
+        return(invisible(
+            request_response("replace_text_in_current_selection",
+                text = location,
+                id = id
+            )
+        ))
+    } else if (missing(location)) {
+        ## handling insertText(text = "text")
+        return(invisible(request_response("replace_text_in_current_selection",
+            text = text,
+            id = id
+        )))
+    } else if (is.null(location) && missing(text)) {
+        ## handling insertText(NULL)
+        return(invisible(NULL))
+    }
+
+    ## ensure normalised_location is a list containing a possible mix of
+    ## document_position and document_range objects
+    normalised_location <- normalise_pos_or_range_arg(location)
+    normalised_text <- normalise_text_arg(text, length(normalised_location))
+    ## Having normalised we are guaranteed these are the same length.
+    ## Package up all the edits in a query to send to VSCode in an object
+    ## This is done so the edits can be applied in a single edit object, which
+    ## is hopefull closest to RStudio behaviour.
+    query <-
+        mapply(function(location, text) {
+            list(
+                operation = if (rstudioapi::is.document_range(location)) {
+                    "modifyRange"
+                } else {
+                    "insertText"
+                },
+                location = location,
+                text = text
+            )
+        },
+        normalised_location,
+        normalised_text,
+        SIMPLIFY = FALSE
+        )
+
+    invisible(
+        request_response("insert_or_modify_text", query = query, id = id)
+    )
+}
+
+modifyRange <- insertText
+
+readPreference <- function(name, default) {
+    ## in future we could map some rstudio preferences to vscode settings.
+    ## since the caller must provide a default this should work.
+    default
+}
+
+readRStudioPreference <- readPreference
+
+hasFun <- function(name, version_needed = NULL, ...) {
+    if (!is.null(version_needed)) {
+        return(FALSE)
+    }
+
+    exists(x = name, envir = as.environment(rstudio_vsc_mapping), ...)
```

Comment:
Should find directly rather than in non-existing `rstudio_vsc_mapping`?

Reply:
doh.... yeah so this was another useful feature of the explicit mapping. I was only reporting true for things that had a valid implementation

Reply:
I think this is a useful feature. To get it back I'll need to move all the not implemented functions to a separate file which gets sourced into a different environment again. I am really questioning the value of this change now.

Reply:
Yes, the explicit mapping could be useful here. But there's no need to make things complicated. The purpose of this is to make the code easier to maintain, therefore a simple fix for this could be

```r
.vsc.rstudioapi_env <- environment()

hasFun <- function(name, version_needed = NULL, ...) {
    if (!is.null(version_needed)) {
        return(FALSE)
    }

    obj <- .vsc.rstudioapi_env[[name]]
    is.function(obj) && !identical(obj, .vsc_not_yet_implemented)
}

getFun <- function(name, version_needed = NULL, ...) {
    if (!is.null(version_needed)) {
        stop("VSCode does not support used of 'version_needed'.")
    }

    get(name, .vsc.rstudioapi_env, mode = "function", inherits = FALSE)
}
```

Reply:
Yeah okay that's not too bad. :+1:

Reply:
`getFun` needs to call `hasFun` also, and error if `False`, since it's supposed to error if it cannot get the function.

Reply:
Yes, sorry for missing that `getFun` only works with mapped functions.

Reply:
all good that comment was more of a reminder to myself. Looks good.